tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java

```diff
@@ -0,0 +1,158 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+package org.tensorflow.op;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+/**
+ * A class to manage scoped (hierarchical) names for operators.
+ *
+ * <p>{@code NameScope} manages hierarchical names where each component in the hierarchy is
+ * separated by a forward slash {@code '/'}. For instance, {@code nn/Const_72} or {@code
+ * nn/gradient/assign/init}. Each scope is a subtree in this hierarchy.
+ *
+ * <p>Use {@code NameScope} to group related operations within a hierarchy, which for example lets
+ * tensorboard coalesce nodes for better graph visualizations.
+ *
+ * <p>This class is package private, user code creates {@link Scope} which internally delegates
+ * calls to an underlying {@code NameScope}.
+ *
+ * <p>This class is thread-safe.
```

Comment:
If graph construction time seems a non-issue, we can close off the option entirely as you suggest.

One thought: should you be considering more of a wait-and-see approach - it may be less disruptive to improve a less performant but correct implementation than change it from "not thread-safe" to "thread-safe." The reason is that user-side code will likely need to restructure itself to take advantage in the latter case; but in the former case it will improve directly as the framework code improves.

For what it's worth, the locks here are per scope-instance; so to the extent different threads use different scopes there should be no contention. Aiui, possibly the place to watch out may be  [the per-graph global lock](https://github.com/tensorflow/tensorflow/blob/ce02c770fb269e5e607da459bde4f580ef108137/tensorflow/c/c_api_internal.h#L67) in the underlying c_api abstraction that guards operation creation itself.